### PR TITLE
feat: verify tx cycles in relay protocol

### PR DESCRIPTION
**Pull request by jjyr**

### CODE REVIEW

File: rpc/src/module/trace.rs [hunk 1]
```diff
@@ -31,22 +35,47 @@ impl<CI: ChainIndex + 'static> TraceRpc for TraceRpcImpl<CI> {
...
 
-        let fbb = &mut FlatBufferBuilder::new();
-        let message = RelayMessage::build_transaction(fbb, &tx);
-        fbb.finish(message, None);
+        match tx_result {
+            Ok(cycles) => {
+                let entry = PoolEntry::new(tx.clone(), 0, Some(cycles));
+                self.shared
+                    .chain_state()
+                    .lock()
```
Comment (zhangsoledad) [inline rpc/src/module/trace.rs]: lock acquired twice

File: sync/src/tests/relayer.rs [hunk 2]
```diff
@@ -212,8 +217,16 @@ fn relay_compact_block_with_missing_indexs() {
                 .collect::<Vec<_>>();
 
             [3, 5].iter().for_each(|i| {
+                let tx = &txs[*i];
+                let cycles = {
+                    let chain_state = shared1.chain_state().lock();
```
Comment (zhangsoledad) [inline sync/src/tests/relayer.rs]: lock acquired in loop

File: rpc/src/module/trace.rs (hunk 1 above)
Comment (jjyr) [reply rpc/src/module/trace.rs]: The intention is to release the lock before  broad casting

File: sync/src/tests/relayer.rs (hunk 2 above)
Comment (jjyr) [reply sync/src/tests/relayer.rs]: Its tests code, hold the lock outside will block the tests(the node is in the same process).

File: rpc/src/module/trace.rs (hunk 1 above)
Comment (zhangsoledad) [reply rpc/src/module/trace.rs]:
```

**Comment by doitian** [comment]
bors try

**Comment by doitian** [comment]
It seems to break the integration test https://github.com/nervosnetwork/ckb/runs/80682865

How to run integration test locally:

```
cargo build
cargo run -p ckb-test target/debug/ckb
```

cc @quake

**Comment by jjyr** [comment]
bors try

**Comment by jjyr** [comment]
bors try

File: sync/src/relayer/transaction_process.rs [hunk 3]
```diff
@@ -32,31 +38,48 @@ where
...
-            let max_block_cycles = self.relayer.shared.consensus().max_block_cycles();
-            chain_state.add_tx_to_pool(tx.clone(), max_block_cycles)
-        };
-
-        if ret.is_ok() {
-            let fbb = &mut FlatBufferBuilder::new();
-            let message = RelayMessage::build_transaction(fbb, &tx);
-            fbb.finish(message, None);
+        let (tx, relay_cycles): (Transaction, Cycle) = (*self.message).into();
+        let chain_state = self.relayer.shared.chain_state().lock();
```
Comment (zhangsoledad) [inline sync/src/relayer/transaction_process.rs]: original code shrink lock-acquisition here

Comment (jjyr) [reply sync/src/relayer/transaction_process.rs]: Fixed

**Pull request merged**